### DISCUSSION TITLE
Debian package

### DISCUSSION THREAD

**Discussion by mobin-2008** [davmac314/dinit#149] [Show and tell]
Hi! I created Dinit package for debian. It's available at [mentors.debian.net](https://mentors.debian.net/package/dinit/) and I looking for a sponsor for pushing it to Debian unstable repo. Feel free to send comments and reviews in mentors.

Regrads

**Comment by davmac314** [comment]
One obvious thing: could it be upgraded to 0.16.1? That has a few bugs fixed compared to 0.16.0

**Reply by mobin-2008** [reply]
I'll fix this thing. (Working with Debian interfaces is need more time)

**Comment by mobin-2008** [comment]
Dinit 0.16.1 is now available in [Debian mentors](https://mentors.debian.net).

**Comment by mobin-2008** [comment]
Seems like no one interest in moving dinit to debian official repos. I will try again when stable version of Dinit released (or at least beta release).

**Comment by ProwlerGr** [comment]
It is becoming a popular option in antiX:
https://www.antixforum.com/forums/topic/antix-23-2-init-diversity-2025-remaster-edition/

I maintain an independent repo here
https://gitlab.com/init-diversity/dinit

**Reply by mobin-2008** [reply]
Thanks!
https://gitlab.com/init-diversity/dinit/dinit-init-diversity looks good but one suggestion.

In https://gitlab.com/init-diversity/dinit/dinit-init-diversity/-/blob/init-diversity/debian/rules there is a override for configure script which is for early days of configure script. As of now configure supports well-known CXXFLAGS, LDFLAGS, CPPFLAGS (both from command-line and environment) so there is no need for `override_dh_auto_configure` anymore.

**Reply by ProwlerGr** [reply]
Maybe give it a spin & post your impressions in the antiX forum.
I will also opt to create multi-init spins for MX Devuan & Debian.

**Reply by mobin-2008** [reply]
Sure but I don't have a Devuan/AntiX system to actually test your work right now. I will post my feedback when I tested it.

**Reply by davmac314** [reply]
From that thread, I find this troubling:

> After troubleshooting commands seem to need to be:
> reboot: /usr/lib/dinit/shutdown -r --system

You should not be running shutdown with the "--system" argument; that's going to bypass service shutdown.

If dinit really core dumps when a normal shutdown is attempted, I'd love to know why. Did you build with `SBINDIR=/usr/lib/dinit` as would be required if dinit's shutdown is in that location?

**Reply by ProwlerGr** [reply]
Hi Davin,
Yes I worked out what the problem was (It was an issue with my wrapper multi-init script - nothing wrong with dinit's native function).

A link to the scripts I use here [https://gitlab.com/init-diversity/general/init-diversity-tools/-/tree/debian/init-diversity-tools?ref_type=heads](url). 
You are also very welcome to post in the antiX forum, we would love to see you engaging with the there.
Keep up the good work!

**Reply by ProwlerGr** [reply]
> Did you build with SBINDIR=/usr/lib/dinit as would be required if dinit's shutdown is in that location?

Something like that, we have diverted all dinit binaries to a folder under /lib/dinit & added some symlinks from /usr/lib/dinit back to /usr/sbin. As we have multiple inits in a single installation the shutdown - poweroff - reboot have been replaced with a wrapper as per the above post.

Maybe try the spin in a spare partition or a VM, I'm sure that you will find it quite interesting as an init developer...

**Comment by ProwlerGr** [comment]
@mobin-2008 @davmac314 , just a heads up, there is an official beta of antiX-25 which has dinit included
Link to thread:
https://www.antixforum.com/forums/topic/antix-25-full-beta1available-for-public-testing

Your testing and feedback would be highly appreciated.

**Reply by davmac314** [reply]
Unfortunate timing, in that I intend to release 0.20 (final beta, "really this time") shortly. But that doesn't matter - thanks for the heads-up!

**Reply by ProwlerGr** [reply]
This is just the official beta, it may still make it into the final official release if it comes soonish & builds & plays well.

**Reply by mobin-2008** [reply]
@ProwlerGr I tried the testing iso, and it worked just fine as a live image on my virtual machine. Some suggestions to make it even better:

1. Services are based on my old, abandoned project [Dinit-on-Devuan](https://github.com/Dinit-on-Devuan/dinitscripts), and I'm happy it is useful for other users for booting Debian-based systems. However, time has changed since I started that project. Chimera Linux introduced a new style ([dinit-chimera](https://github.com/chimera-linux/dinit-chimera)) of system-bootup scripts and target system, which is used by Chimera Linux, Artix Linux, and eweOS. It would be really good if services get an overhaul based on dinit-chimera targets at least, or even better, porting/patching dinit-chimera for Debian-based systems, which provides additional functionalities (with extra binaries) in comparison to other init systems.

2. Some services (like `sshd` and `dbus`) use a shell script for initial preparation purposes, but I think there is an oversight about them; the shell script doesn't execute into the daemon binary, and instead it just forks the daemon. This is bad because it defeats the purpose of supervision provided by Dinit (i.e. Dinit is watching `/bin/sh /etc/dinit.d/scripts/sshd` instead of the actual `sshd` daemon). The fix is to change this:
```
# The final daemon launching
/usr/sbin/sshd -D -e -f /etc/ssh/sshd_config
```
to this
```
# The final daemon launching
exec /usr/sbin/sshd -D -e -f /etc/ssh/sshd_config
```

3. It would be really awesome to add support for turnstile for all init systems (not only Dinit) to provide easy-to-use user services. See https://github.com/chimera-linux/turnstile for the turnstile project. It already supports runit and Dinit, and support for other init systems can be added. It is available in Aritx Linux and Void Linux.

Thank Antix people for adding Dinit as an init option. Best wishes to them :)

**Reply by davmac314** [reply]
> Some services (like sshd and dbus) use a shell script for initial preparation purposes ...The fix is to change this:

In general service processes should not need to be executed via a shell script at all. If there is preparatory work it can be separated to another (scripted) service and the main service can depend on the preparation service. This gives better feedback in case of failure; it is clear whether preparation failed, or the service itself. It also can prevent pointless attempts to restart the service if the preparation script fails.

**Reply by ProwlerGr** [reply]
Thanks for the feedback & the tips guys, it would be great to see you in our forum.

Getting the various service files working has been our top priority - optimising them will be the sure next step.
Will certainly look into turnstile in the future, antiX currently doesn't use a login tracker (just seatd).
We have just started looking at user services & found a workaround (suitable to the current state of antiX & possibly MX) to launch them reliably at user log-in, but I would really like to see if turnstille can be usefull with the s6 implementations as well.

**Reply by ProwlerGr** [reply]
@mobin-2008 @davmac314 I had to make some changes to our connmand service (to address an issue related to a broken resolv.conf), but I thought to share this as a good example of a daemon service requiring preparatory actions before launching
https://gitlab.com/init-diversity/dinit/dinit-service-connmand/-/blob/debian/services/connmand?ref_type=heads

> type = scripted
command = sh -c "rm -rf /etc/resolv.conf && exec /usr/sbin/connmand -n --nobacktrace --nodnsproxy"
logfile	= /tmp/connmand.log
> depends-on = dbus

As you can see this now includes the "exec" command but it is in a shell wrapper. Is this the recommended approach?

**Reply by davmac314** [reply]
> Is this the recommended approach?

As I said above, my recommended approach is to separate the preparatory work into a separate (scripted) service that the main service depends on. In this example you would have the secondary service (`connmand-prep` or whatever you like to call it) perform the `rm -rf ...` action (since it's only one command, no need for a script) and the main service executes the `connmand` command.

The only drawback to that approach is that it won't do the prep work if the service is restarted. If you really need to do that, then either execute both commands in the one service via a `sh -c` command as in your example, or use an actual shell script on the filesystem and have it `exec` the daemon as its final action.

**Reply by davmac314** [reply]
(I'd like to add: it's a bit sad if `connmand` needs `/etc/resolv.conf` removed but can't do that itself!).

**Reply by ProwlerGr** [reply]
Thanks Davin,
Just FYI, connman itself can remove & create resolv.conf (it creates it in its /run directory & symlinks it to /etc/resolv.conf) The problem occurs when something abnormal happens & connman does not exit properly it leaves a broken symlink at /etc/resolv.conf. It does not recognise the broken symlink & does not replace it with a healthy file (resulting in no dns resolution - no internet for users). This workaround works like a charm ;)

**Reply by ProwlerGr** [reply]
Just a heads up, we succesfully built dinit tag version 0.20.0 & it is already in the antix Trixie repo's 
https://repo.antixlinux.com/trixie/pool/main/d/dinit/
There will be a new beta2 release soonish that is intended to have this latest version included

**Comment by yosh781** [comment]
Thank you for your work. I set up Dinit to work on Devuan KDE with Turnstile,  thanks to your work in AntiX. 
https://i.postimg.cc/DyCZ37YM/Snimok-ekrana-20260419-194809.png

**Reply by ProwlerGr** [reply]
Thanks, I'll give it a test drive when I find some time...
